packages/edit-post/src/components/header/header-toolbar/index.js

```diff
@@ -73,6 +76,10 @@ function HeaderToolbar() {
 	/* translators: accessibility text for the editor toolbar */
 	const toolbarAriaLabel = __( 'Document tools' );
 
+	const toggleListView = useCallback(
+		() => setIsListViewOpened( ! isListViewOpen ),
+		[ setIsListViewOpened, isListViewOpen ]
```

Comment:
The setter and value here need to be dependencies since this is using a data store.

As an aside this is avoidable in similar handlers when using useState. (Eg where a state setter [does not change](https://reactjs.org/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often) and using a [functional update](https://reactjs.org/docs/hooks-reference.html#functional-updates) to avoid a closure with stale data. )

Reply:
Thanks for finding all these unnecessary renders!

Yes, the difference in ergonomics between `useState` entities and `useSelect/useDispatch` ones is striking. However, as of #31078 we encourage an "advanced" use case of `useSelect` that fits the current case very well:

```diff
diff --git a/packages/edit-post/src/components/header/header-toolbar/index.js b/packages/edit-post/src/components/header/header-toolbar/index.js
index f3e16a6f2b..1ddab499e2 100644
--- a/packages/edit-post/src/components/header/header-toolbar/index.js
+++ b/packages/edit-post/src/components/header/header-toolbar/index.js
@@ -77,10 +77,10 @@ function HeaderToolbar() {
 	/* translators: accessibility text for the editor toolbar */
 	const toolbarAriaLabel = __( 'Document tools' );
 
-	const toggleListView = useCallback(
-		() => setIsListViewOpened( ! isListViewOpen ),
-		[ setIsListViewOpened, isListViewOpen ]
-	);
+	const { isListViewOpened } = useSelect( editPostStore );
+	const toggleListView = useCallback( () => {
+		setIsListViewOpened( ! isListViewOpened() );
+	}, [] );
 	const overflowItems = (
 		<>
 			<ToolbarItem
diff --git a/packages/edit-site/src/components/header/index.js b/packages/edit-site/src/components/header/index.js
index 2d03cf3e57..e930c45f4b 100644
--- a/packages/edit-site/src/components/header/index.js
+++ b/packages/edit-site/src/components/header/index.js
@@ -99,10 +99,10 @@ export default function Header( {
 		}
 	}, [ isInserterOpen, setIsInserterOpened ] );
 
-	const toggleListView = useCallback(
-		() => setIsListViewOpened( ! isListViewOpen ),
-		[ setIsListViewOpened, isListViewOpen ]
-	);
+	const { isListViewOpened } = useSelect( editSiteStore );
+	const toggleListView = useCallback( () => {
+		setIsListViewOpened( ! isListViewOpened() );
+	}, [] );
 
 	return (
 		<div className="edit-site-header">
```

I personally took the liberty of also removing the setter from the dependencies because I don't think that object will ever change, but that's up to debate. What really matters is that the callback itself no longer needs to be recomputed based on the change of a piece of state that the callback only cares about in an eventual future.

It's not as nice as `setFoo( foo => ! foo )`, but it's a start.

Reply:
Nice, thanks @mcsf !